zerver/lib/bugdown/__init__.py

```diff
@@ -1248,64 +1248,68 @@ def run(self, lines: List[Text]) -> List[Text]:
                 inserts += 1
         return copy
 
-class AutoNumberOListPreprocessor(markdown.preprocessors.Preprocessor):
-    """ Finds a sequence of lines numbered by the same number"""
-    RE = re.compile(r'^([ ]*)(\d+)\.[ ]+(.*)')
-    TAB_LENGTH = 2
+class ParagraphProcessor(markdown.blockprocessors.ParagraphProcessor):
+    """ Before a paragraph is inserted, look for lines that form
+        ordered list with the same number on each line and change
+        them to incrementing ones."""
 
-    def run(self, lines):
-        # type: (List[Text]) -> List[Text]
+    RE = re.compile(r'^([ ]*)(\d+)(\.[ ]+.*)')
+
+    def run(self, parent, blocks):
+        # type: (Element, List[Text]) -> None
+
+        if 'blockquote' not in self.parser.state:
+            block = self.renumber_paragraph(blocks.pop(0))
+            blocks.insert(0, block)
+
+        super().run(parent, blocks)
+
+    def renumber_paragraph(self, block):
```

Comment:
Can you explain the logic for how this processor will behavior correctly?  I'm having a hard time reading it.